### PR TITLE
Add async interpolated-string Throws overloads and ValueTask<T> rejection

### LINKED ISSUE #9515: Quality: Async Exception Assertion API Completeness Gaps in Assert.ThrowsAsync / Assert.ThrowsExactlyAsync
### 🎯 Repository Quality Improvement Report — Async Exception Assertion Gaps

**Analysis Date**: 2026-06-29 | **Focus Area**: Async Exception Assertion API Gaps (custom)

### Executive Summary

`Assert.ThrowsAsync<T>` and `Assert.ThrowsExactlyAsync<T>` have only **4 overloads total** (2 per method), while their sync counterparts `Assert.Throws<T>` / `Assert.ThrowsExactly<T>` each have **6** — a 67% gap. Two things are missing:

1. **Interpolated string handler (ISH) overloads** — The sync methods have `Assert.Throws<Ex>(action, $"context: {x}")`, the async variants don't. `MSTEST0064 PreferAsyncAssertion` explicitly short-circuits on `HasInterpolatedStringHandlerParameter(targetMethod)` precisely because no async equivalent exists, leaving ISH users without analyzer guidance.

2. **`Func<ValueTask>` overloads** — MSTest accepts `ValueTask`-returning test methods, and library code increasingly returns `ValueTask`. There's no direct way to pass such a delegate to `ThrowsAsync` without an intermediate `.AsTask()` call.

A tertiary issue: `TryGetTaskFromValueTaskAsync` uses `invokeResult as ValueTask?`, which silently returns `null` for `ValueTask<T>` returns (non-generic `ValueTask<T>` is already rejected at validation time, but the error message is the generic `UTA_ErrorIncorrectTestMethodSignature` rather than a clear "use non-generic ValueTask" hint).

### 📊 API Surface Comparison

| Method | `string` msg | `Func<Ex?,string>` builder | Interpolated handler | `Func<ValueTask>` |
|--------|:-----------:|:--------------------------:|:--------------------:|:-----------------:|
| `Throws<T>(Action, ...)` | ✅ | ✅ | ✅ | n/a |
| `Throws<T>(Func<object?>, ...)` | ✅ | ✅ | ✅ | n/a |
| `ThrowsAsync<T>(Func<Task>, ...)` | ✅ | ✅ | ❌ | ❌ |
| `ThrowsExactlyAsync<T>(Func<Task>, ...)` | ✅ | ✅ | ❌ | ❌ |

---

#### Task 1: Add ISH overloads for `ThrowsAsync` and `ThrowsExactlyAsync`

**Priority**: High | **Effort**: Medium

Add `AssertNonStrictThrowsAsyncInterpolatedStringHandler<TException>` and `AssertThrowsExactlyAsyncInterpolatedStringHandler<TException>` structs in `Assert.ThrowsException.InterpolatedStringHandlers.cs`, and the corresponding public overloads:

```csharp
public static Task<TException> ThrowsAsync<TException>(
    Func<Task> action,
    [InterpolatedStringHandlerArgument(nameof(action))] ref AssertNonStrictThrowsAsyncInterpolatedStringHandler<TException> message,
    [CallerArgumentExpression(nameof(action))] string actionExpression = "")
    where TException : Exception;
// ... same for ThrowsExactlyAsync
```

The async handlers must capture the interpolation lazily (format only on failure, after the awaited task completes). New `PublicAPI.Unshipped.txt` entries required.

---

#### Task 2: Add `Func<ValueTask>` overloads for `ThrowsAsync` and `ThrowsExactlyAsync`

**Priority**: Medium | **Effort**: Small

```csharp
public static Task<TException> ThrowsAsync<TException>(
    Func<ValueTask> action,
    string? message = "",
    [CallerArgumentExpression(nameof(action))] string actionExpression = "")
    where TException : Exception;

public static Task<TException> ThrowsAsync<TException>(
    Func<ValueTask> action,
    Func<Exception?, string> messageBuilder,
    [CallerArgumentExpression(nameof(action))] string actionExpression = "")
    where TException : Exception;
// ... same for ThrowsExactlyAsync; add ISH variants if Task 1 is done first
```

Implement by converting `Func<ValueTask>` to `Func<Task>` via `() => action().AsTask()` and delegating to the existing core. `PublicAPI.Unshipped.txt` entries required.

---

#### Task 3: Update MSTEST0064 to cover ISH case once Task 1 ships

**Priority**: High (depends on Task 1) | **Effort**: Small

Remove the `HasInterpolatedStringHandlerParameter` short-circuit from `PreferAsyncAssertionAnalyzer.cs` and update `PreferAsyncAssertionFixer.cs` to correctly rewrite `Assert.Throws(action, $"...")` → `await Assert.ThrowsAsync(action, $"...")` including handler-type substitution. Add unit tests for the new fix path.

---

#### Task 4: Add explicit `ValueTask<T>` rejection message in test method validation

**Priority**: Low | **Effort**: Small

In `TestMethodValidator.cs`, detect `ValueTask<T>` (generic) specifically and emit a targeted resource message — e.g. `"ValueTask<T> is not a valid test method return type; use non-generic ValueTask instead."` — rather than the generic `UTA_ErrorIncorrectTestMethodSignature`. Consider surfacing this via MSTEST0003 (`TestMethodShouldBeValid`).

---

#### Task 5: Add tests for new async assertion overloads

**Priority**: High (prerequisite for Tasks 1–2) | **Effort**: Medium

In `test/UnitTests/TestFramework.UnitTests`:
- Unit tests for `ThrowsAsync(Func<Task>, [ISH])` verifying lazy interpolation (string not evaluated on success).
- Unit tests for `ThrowsAsync(Func<ValueTask>, ...)` covering pass/fail/wrong-exception scenarios.
- Acceptance test verifying MSTEST0064 now fires and fixes `Assert.Throws(action, $"...")` in async test methods.

---

### 📊 Historical Context

<details>
<summary><b>Previous Focus Areas (last 5)</b></summary>

| Date | Focus Area | Type |
|------|------------|------|
| 2026-06-26 | analyzer-performance-category-deficit | Custom |
| 2026-06-25 | treenode-filter-completeness | Custom |
| 2026-06-24 | itestdatasource-extensibility-quality | Custom |
| 2026-06-23 | testconfig-json-schema-and-discoverability | Custom |
| 2026-06-22 | publicapi-versioning-hygiene | Custom |

</details>

*Next analysis: 2026-06-30 — Focus area selected based on diversity algorithm*

> 🤖 **Automated content by GitHub Copilot.** Posted via a maintainer's GitHub token, so it appears under their account — the account owner did **not** write or approve this content personally. Generated by the [Repository Quality Improver](https://github.com/microsoft/testfx/actions/runs/28407118437/agentic_workflow) workflow. · 1.3K AIC · ⌖ 14.1 AIC · ⊞ 45.8K · [◷]( · [◷](https://github.com/search?q=repo%3Amicrosoft%2Ftestfx+is%3Aissue+%22gh-aw-workflow-call-id%3A+microsoft%2Ftestfx%2Frepository-quality-improver%22&type=issues))
>
<details>
<summary>Add this agentic workflows to your repo</summary>

To install this agentic workflow, run

```
gh aw add githubnext/agentics/workflows/repository-quality-improver.md@main
```
</details>

> - [x] expires  on Jul 1, 2026, 10:48 PM UTC

### PR DESCRIPTION
**Pull request by Evangelink**
Fixes #9515.

Closes the async exception-assertion API gaps between `Assert.ThrowsAsync`/`ThrowsExactlyAsync` and their sync counterparts.

### Task 1 — Async interpolated-string-handler (ISH) overloads (High)
- Added `AssertNonStrictThrowsAsyncInterpolatedStringHandler<TException>` and `AssertThrowsExactlyAsyncInterpolatedStringHandler<TException>` structs.
- Added `Assert.ThrowsAsync`/`ThrowsExactlyAsync(Func<Task>, ref handler, …)` overloads. The interpolated message is built **lazily** — only when the awaited assertion actually fails.
- New `PublicAPI.Unshipped.txt` entries (base + `net`).

### Task 3 — MSTEST0064 now covers the ISH case (High)
- Removed the `HasInterpolatedStringHandlerParameter` short-circuit in `PreferAsyncAssertionAnalyzer`, so `Assert.Throws(action, $"...")` in async-capable test methods now reports MSTEST0064 and is auto-fixed to `await Assert.ThrowsAsync(action, $"...")` (the existing fixer handles handler-type substitution).

### Task 4 — Explicit `ValueTask<T>` rejection (Low)
- `TestMethodValidator` now detects generic `ValueTask<T>` return types and emits a targeted message (`UTA_ErrorGenericValueTaskReturnType`) telling the user to use non-generic `ValueTask`, instead of the generic `UTA_ErrorIncorrectTestMethodSignature`. Resource added and `.xlf` files regenerated via `UpdateXlf`.

### Task 5 — Tests (High)
- TestFramework unit tests verifying lazy interpolation (message not evaluated on success; rendered on failure) for both `ThrowsAsync` and `ThrowsExactlyAsync`.
- Analyzer code-fix tests for the new MSTEST0064 ISH fix path (both `Throws` and `ThrowsExactly`).
- Validator tests for `ValueTask` (valid) and `ValueTask<T>` (rejected with the targeted message).

All affected suites pass locally: analyzer 28/28, TestFramework async 13/13, validator 15/15.

## Task 2 (`Func<ValueTask>` overloads) — intentionally not implemented

Adding `Func<ValueTask>` overloads alongside the existing `Func<Task>` ones makes the most common assertion patterns **ambiguous and source-breaking**:

```csharp
Assert.ThrowsAsync<MyException>(() => throw new MyException());      // CS0121
Assert.ThrowsAsync<MyException>(async () => await DoAsync());        // CS0121
```

A bare `() => throw ...` or `async () => ...` lambda matches both `Func<Task>` and `Func<ValueTask>` equally well, so the compiler can no longer pick an overload — this breaks existing call sites (including MSTest's own tests). The sync API avoids this only because one overload takes `Action` (void-returning), which has no async analogue. There is no overload arrangement that adds `Func<ValueTask>` without this break, so it was deferred rather than ship a breaking change to the flagship assertion API. Happy to revisit for a major version.

> Generated with the help of GitHub Copilot.

### CODE REVIEW

**Comment by Evangelink** [comment]
### 🧪 Test quality grade — PR #9532

14 new test methods identified across 3 files. All 14 earn **A (90–100)**. The suite is thorough: each new API overload (`messageBuilder` and ISH for `ThrowsAsync`/`ThrowsExactlyAsync`, `ValueTask<T>` rejection) has coverage for the happy path, every failure mode, and — for the ISH overloads — lazy-evaluation semantics. No issues found.

<table>
  <thead>
    <tr><th>Δ</th><th>Test</th><th>Grade</th><th>Band</th><th>Notes</th></tr>
  </thead>
  <tbody>
    <tr>
      <td>new</td>
      <td><code>AssertTests.<br>ThrowsAsync_<br>WithInterpolation_<br>DoesNotEvaluateMessageOnSuccess</code></td>
      <td>A</td>
      <td>90–100</td>
      <td>Clear AAA; local-function marker verifies the ISH message is not evaluated on success.</td>
    </tr>
    <tr>
      <td>new</td>
      <td><code>AssertTests.<br>ThrowsAsync_<br>WithInterpolation_<br>RendersMessageOnFailure</code></td>
      <td>A</td>
      <td>90–100</td>
      <td>Verifies exact failure-message rendering including the interpolated value.</td>
    </tr>
    <tr>
      <td>new</td>
      <td><code>AssertTests.<br>ThrowsAsync_<br>WithMessageBuilder_<br>FailsBecauseNoException</code></td>
      <td>A</td>
      <td>90–100</td>
      <td>Multiple assertions verify builder invocation, null exception arg, and exact error message.</td>
    </tr>
    <tr>
      <td>new</td>
      <td><code>AssertTests.<br>ThrowsAsync_<br>WithMessageBuilder_<br>FailsBecauseTypeMismatch</code></td>
      <td>A</td>
      <td>90–100</td>
      <td>Verifies builder invocation, correct exception reference passed, and exact error message content.</td>
    </tr>
    <tr>
      <td>new</td>
      <td><code>AssertTests.<br>ThrowsAsync_<br>WithMessageBuilder_<br>Passes</code></td>
      <td>A</td>
      <td>90–100</td>
      <td>Verifies the messageBuilder is not called when the assertion passes (lazy evaluation).</td>
    </tr>
    <tr>
      <td>new</td>
      <td><code>AssertTests.<br>ThrowsExactlyAsync_<br>WithInterpolation_<br>DoesNotEvaluateMessageOnSuccess</code></td>
      <td>A</td>
      <td>90–100</td>
      <td>No issues found.</td>
    </tr>
    <tr>
      <td>new</td>
      <td><code>AssertTests.<br>ThrowsExactlyAsync_<br>WithInterpolation_<br>RendersMessageOnFailure</code></td>
      <td>A</td>
      <td>90–100</td>
      <td>No issues found.</td>
    </tr>
    <tr>
      <td>new</td>
      <td><code>AssertTests.<br>ThrowsExactlyAsync_<br>WithMessageBuilder_<br>FailsBecauseNoException</code></td>
      <td>A</td>
      <td>90–100</td>
      <td>No issues found.</td>
    </tr>
    <tr>
      <td>new</td>
      <td><code>AssertTests.<br>ThrowsExactlyAsync_<br>WithMessageBuilder_<br>FailsBecauseTypeMismatch</code></td>
      <td>A</td>
      <td>90–100</td>
      <td>Verifies builder invocation, exception type passed to builder, and exact error message.</td>
    </tr>
    <tr>
      <td>new</td>
      <td><code>AssertTests.<br>ThrowsExactlyAsync_<br>WithMessageBuilder_<br>Passes</code></td>
      <td>A</td>
      <td>90–100</td>
      <td>No issues found.</td>
    </tr>
    <tr>
      <td>new</td>
      <td><code>PreferAsyncAssertionAnalyzerTests.<br>WhenAssertionUsesInterpolatedStringHandlerOverload_<br>CodeFixUsesAsyncAssertion</code></td>
      <td>A</td>
      <td>90–100</td>
      <td>Code-fix verification covers diagnostic and transformation for the ISH overload path.</td>
    </tr>
    <tr>
      <td>new</td>
      <td><code>PreferAsyncAssertionAnalyzerTests.<br>WhenAssertThrowsUsesInterpolatedStringHandlerOverload_<br>CodeFixUsesAsyncAssertion</code></td>
      <td>A</td>
      <td>90–100</td>
      <td>No issues found.</td>
    </tr>
    <tr>
      <td>new</td>
      <td><code>TestMethodValidatorTests.<br>IsValidTestMethodShouldReturnFalseAndReportTargetedWarningForGenericValueTaskReturnType</code></td>
      <td>A</td>
      <td>90–100</td>
      <td>Three assertions cover return value, warning count, and message content — thorough.</td>
    </tr>
    <tr>
      <td>new</td>
      <td><code>TestMethodValidatorTests.<br>IsValidTestMethodShouldReturnTrueForMethodsWithValueTaskReturnType</code></td>
      <td>A</td>
      <td>90–100</td>
      <td>Clear AAA; verifies return value and absence of warnings.</td>
    </tr>
  </tbody>
</table>

<sub>This advisory comment was generated automatically. Grades are heuristic and informational — they do not block merging. Re-run with `/grade-tests`.</sub>



> 🤖 **Automated content by GitHub Copilot.** Posted via a maintainer's GitHub token, so it appears under their account — the account owner did **not** write or approve this content personally. Generated by the [Grade Tests on PR (on open / sync)](https://github.com/microsoft/testfx/actions/runs/28429231578/agentic_workflow) workflow. · 716.9 AIC · ⌖ 32.1 AIC · ⊞ 43.8K · [◷]( · [◷](https://github.com/search?q=repo%3Amicrosoft%2Ftestfx+%22gh-aw-workflow-call-id%3A+microsoft%2Ftestfx%2Fgrade-tests-on-pr.agent%22&type=issues))

File: src/TestFramework/TestFramework/Assertions/Assert.ThrowsException.InterpolatedStringHandlers.cs [hunk 1]
```diff
@@ -139,4 +139,114 @@ internal TException ComputeAssertion(string actionExpression)
...
+        /// <param name="formattedCount">The number of interpolation expressions in the interpolated string.</param>
+        /// <param name="action">The delegate being asserted; the message is only computed when the assertion fails.</param>
+        /// <param name="shouldAppend">When this method returns, indicates whether the interpolated string should be evaluated.</param>
+        public AssertNonStrictThrowsAsyncInterpolatedStringHandler(int literalLength, int formattedCount, Func<Task> action, out bool shouldAppend)
+        {
+            // The interpolated string handler pattern requires deciding 'shouldAppend' synchronously in the
+            // constructor (the compiler skips evaluating the interpolation holes entirely when it is false).
+            // Determining whether the assertion fails requires running the action, so the awaited task is
+            // observed synchronously here. The async core uses ConfigureAwait(false) throughout.
+            _state = IsThrowsAsyncFailingAsync<TException>(action, isStrictType: false).GetAwaiter().GetResult();
```
Comment (Evangelink) [inline src/TestFramework/TestFramework/Assertions/Assert.ThrowsException.InterpolatedStringHandlers.cs]: **[Design / Deadlock risk — MEDIUM]** The `.GetAwaiter().GetResult()` call synchronously blocks the calling thread until the action completes. The code comment correctly notes that `ConfigureAwait(false)` is used throughout `IsThrowsAsyncFailingAsync`, which avoids the most common deadlock on typical ASP.NET-Classic / WPF style synchronization contexts. However, for completeness consider documenting the remaining risk:

```csharp
// WARNING: If the action schedules continuations on the thread being blocked (e.g. inside a
// WindowsFormsSynchronizationContext or DispatcherSynchronizationContext whose pump is not
// re-entrant), this will deadlock. Test environments running on ThreadPool threads are safe.
```

A safer alternative when a non-null `SynchronizationContext` is present:
```csharp
_state = Task.Run(() => IsThrowsAsyncFailingAsync<TException>(action, isStrictType: false)).GetAwaiter().GetResult();
```
`Task.Run` schedules the work on the thread pool, detaching it from any ambient context. The same applies to line 229 for `AssertThrowsExactlyAsyncInterpolatedStringHandler`.

File: src/TestFramework/TestFramework/Assertions/Assert.ThrowsException.InterpolatedStringHandlers.cs [hunk 2]
```diff
@@ -139,4 +139,114 @@ internal TException ComputeAssertion(string actionExpression)
...
+        /// <param name="formattedCount">The number of interpolation expressions in the interpolated string.</param>
+        /// <param name="action">The delegate being asserted; the message is only computed when the assertion fails.</param>
+        /// <param name="shouldAppend">When this method returns, indicates whether the interpolated string should be evaluated.</param>
+        public AssertThrowsExactlyAsyncInterpolatedStringHandler(int literalLength, int formattedCount, Func<Task> action, out bool shouldAppend)
+        {
+            // The interpolated string handler pattern requires deciding 'shouldAppend' synchronously in the
+            // constructor (the compiler skips evaluating the interpolation holes entirely when it is false).
+            // Determining whether the assertion fails requires running the action, so the awaited task is
+            // observed synchronously here. The async core uses ConfigureAwait(false) throughout.
+            _state = IsThrowsAsyncFailingAsync<TException>(action, isStrictType: true).GetAwaiter().GetResult();
```
Comment (Evangelink) [inline src/TestFramework/TestFramework/Assertions/Assert.ThrowsException.InterpolatedStringHandlers.cs]: Same deadlock concern applies here — duplicate of the note on line 174. Either add the `Task.Run` guard or at minimum document the SynchronizationContext caveat in the XML summary so callers know this constructor is not safe when called on a UI thread or other pump-based context.

File: src/TestFramework/TestFramework/Assertions/Assert.ThrowsException.cs [hunk 3]
```diff
@@ -299,4 +299,24 @@ public static Task<TException> ThrowsAsync<TException>(Func<Task> action, Func<E
...
         => ThrowsExceptionAsync<TException>(action, isStrictType: true, messageBuilder, actionExpression);
+
+    /// <inheritdoc cref="ThrowsAsync{TException}(System.Func{System.Threading.Tasks.Task}, string, string)" />
+#pragma warning disable IDE0060 // Remove unused parameter - https://github.com/dotnet/roslyn/issues/76578
+    public static Task<TException> ThrowsAsync<TException>(Func<Task> action, [InterpolatedStringHandlerArgument(nameof(action))] ref AssertNonStrictThrowsAsyncInterpolatedStringHandler<TException> message, [CallerArgumentExpression(nameof(action))] string actionExpression = "")
+#pragma warning restore IDE0060 // Remove unused parameter
+        where TException : Exception
+    {
+        TelemetryCollector.TrackAssertionCall("Assert.ThrowsAsync");
+        return Task.FromResult(message.ComputeAssertion(actionExpression));
```
Comment (Evangelink) [inline src/TestFramework/TestFramework/Assertions/Assert.ThrowsException.cs:314]: **[Behavioral inconsistency with other `ThrowsAsync` overloads — LOW]** Every other `ThrowsAsync` overload (the `string` and `Func<Exception?,string>` ones) returns a *faulted* `Task<TException>` when the assertion fails — the exception is embedded in the task and surfaced only when `await`-ed. Here, if `ComputeAssertion` throws (no enclosing `AssertScope`), the exception propagates **synchronously** from `ThrowsAsync` before `Task.FromResult` ever executes.

For pure `await`-ed usage the two behaviors are equivalent, so this is **not a bug**. It's worth a brief note in the XML summary (or a code comment) so maintainers are not surprised if a future refactor tries to match task-based semantics:

```csharp
// Note: unlike the string-message overloads, assertion failure throws synchronously here
// because the action has already been run in the ISH constructor. For await-only callers
// the observable behaviour is identical.
```

File: test/UnitTests/MSTestAdapter.PlatformServices.UnitTests/Discovery/TestMethodValidatorTests.cs [hunk 4]
```diff
@@ -140,6 +140,29 @@ public void IsValidTestMethodShouldReturnTrueForNonAsyncMethodsWithTaskReturnTyp
...
+        SetupTestMethod();
+        MethodInfo methodInfo = typeof(DummyTestClass).GetMethod(
+            "MethodWithValueTaskReturnType",
+            BindingFlags.Instance | BindingFlags.Public)!;
+
+        _testMethodValidator.IsValidTestMethod(methodInfo, _type, _warnings).Should().BeTrue();
+        _warnings.Count.Should().Be(0);
+    }
+
+    public void IsValidTestMethodShouldReturnFalseAndReportTargetedWarningForGenericValueTaskReturnType()
```
Comment (Evangelink) [inline test/UnitTests/MSTestAdapter.PlatformServices.UnitTests/Discovery/TestMethodValidatorTests.cs:154]: **[Confirmed correct — no issue]** These new test methods (`IsValidTestMethodShouldReturnTrueForMethodsWithValueTaskReturnType` and `IsValidTestMethodShouldReturnFalseAndReportTargetedWarningForGenericValueTaskReturnType`) intentionally have no `[TestMethod]` attribute. The class inherits from `TestContainer` (from `TestFramework.ForTestingMSTest`), where **every public parameterless method** is automatically treated as a test case by the engine — `[TestMethod]` would be redundant. The omission is correct and consistent with all other methods in this file.

File: test/UnitTests/TestFramework.UnitTests/Assertions/AssertTests.ThrowsExceptionTests.cs [hunk 5]
```diff
@@ -346,6 +346,60 @@ message constructed via builder.
...
+            wasMessageEvaluated = true;
+            return "marker";
+        }
+
+        await Assert.ThrowsAsync<ArgumentNullException>(() => Task.FromException(new ArgumentNullException()), $"context: {Marker()}");
+
+        wasMessageEvaluated.Should().BeFalse();
+    }
+
+    public async Task ThrowsAsync_WithInterpolation_RendersMessageOnFailure()
```
Comment (Evangelink) [inline test/UnitTests/TestFramework.UnitTests/Assertions/AssertTests.ThrowsExceptionTests.cs:363]: **[Missing test scenario — LOW]** The four new tests cover the *success* path (correct exception type → message not evaluated) and the *no-exception* failure path. Consider adding a test for the *wrong exception type* failure path, e.g.:

```csharp
public async Task ThrowsAsync_WithInterpolation_RendersMessageOnWrongExceptionType()
{
    Func<Task> action = async () =>
        await Assert.ThrowsAsync<ArgumentNullException>(
            () => Task.FromException(new InvalidOperationException()), $"context: {42}");
    (await action.Should().ThrowAsync<AssertFailedException>())
        .Which.Message.Should().Contain("context: 42");
}
```
This path goes through the `WrongExceptionType` branch of `ReportThrowsFailed` and would guard against a regression where the wrong-exception branch skips the custom message.

File: src/Adapter/MSTestAdapter.PlatformServices/Discovery/TestMethodValidator.cs [hunk 6]
```diff
@@ -74,4 +76,14 @@ internal virtual bool IsValidTestMethod(MethodInfo testMethodInfo, Type type, IC
...
     }
+
+    private static bool IsGenericValueTaskReturnType(MethodInfo testMethodInfo)
+    {
+        Type returnType = testMethodInfo.ReturnType;
+
+        // Compare by name to avoid forcing a load of System.Threading.Tasks.Extensions on platforms where
+        // ValueTask<T> lives in a separate assembly (netstandard2.0 / .NET Framework).
+        return returnType.IsGenericType
+            && returnType.GetGenericTypeDefinition().FullName == "System.Threading.Tasks.ValueTask`1";
```
Comment (Evangelink) [inline src/Adapter/MSTestAdapter.PlatformServices/Discovery/TestMethodValidator.cs:94]: **[Null-safety — informational]** `returnType.GetGenericTypeDefinition().FullName` returns `null` for open generic types constructed in dynamic/emit scenarios where the type has no proper assembly context. For any `MethodInfo` discovered by the normal reflection pipeline from real test assemblies this will never be null, so the `== "..."` comparison (which returns `false` on `null`) is safe. The string literal `"System.Threading.Tasks.ValueTask\`1"` is the canonical CLR name — confirmed correct.

File: src/TestFramework/TestFramework/Assertions/Assert.ThrowsException.InterpolatedStringHandlers.cs (hunk 1 above)
Comment (Evangelink) [reply src/TestFramework/TestFramework/Assertions/Assert.ThrowsException.InterpolatedStringHandlers.cs]: Addressed in 8db561ec7: the synchronous check in both ISH constructors now runs via `Task.Run(() => IsThrowsAsyncFailingAsync(...))` so the action's awaits resume on the thread pool instead of a captured SynchronizationContext that's being blocked by `GetResult()`, removing the deadlock risk. Updated the comments accordingly.

File: src/Adapter/MSTestAdapter.PlatformServices/Discovery/TestMethodValidator.cs [hunk 7]
```diff
@@ -59,7 +59,9 @@ internal virtual bool IsValidTestMethod(MethodInfo testMethodInfo, Type type, IC
 
         if (!isValidTestMethod)
         {
-            string message = string.Format(CultureInfo.CurrentCulture, Resource.UTA_ErrorIncorrectTestMethodSignature, type.FullName, testMethodInfo.Name);
+            string message = IsGenericValueTaskReturnType(testMethodInfo)
+                ? string.Format(CultureInfo.CurrentCulture, Resource.UTA_ErrorGenericValueTaskReturnType, type.FullName, testMethodInfo.Name)
+                : string.Format(CultureInfo.CurrentCulture, Resource.UTA_ErrorIncorrectTestMethodSignature, type.FullName, testMethodInfo.Name);
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply src/Adapter/MSTestAdapter.PlatformServices/Discovery/TestMethodValidator.cs]: Good catch. Addressed in 8db561ec7: the targeted `ValueTask<T>` message is now only emitted when the return type is the actual cause of invalidity (i.e. the method is accessible, non-static and non-abstract). When the method is invalid for another reason it falls back to the generic signature message. Added a validator test covering a static `ValueTask<int>` method to lock this in.

File: src/TestFramework/TestFramework/Assertions/Assert.ThrowsException.cs (hunk 3 above)
Comment (Evangelink) [reply src/TestFramework/TestFramework/Assertions/Assert.ThrowsException.cs:314]: Documented in 8db561ec7: added a comment to both interpolated `ThrowsAsync`/`ThrowsExactlyAsync` overloads noting that assertion failure throws synchronously (the action already ran in the ISH constructor), unlike the faulted-task string/builder overloads, and that the await-only behaviour is identical.

File: src/Adapter/MSTestAdapter.PlatformServices/Discovery/TestMethodValidator.cs (hunk 6 above)
Comment (Evangelink) [reply src/Adapter/MSTestAdapter.PlatformServices/Discovery/TestMethodValidator.cs:94]: Added in 8db561ec7: `ThrowsAsync_WithInterpolation_RendersMessageOnWrongExceptionType` plus an equivalent `ThrowsExactlyAsync` variant, both asserting the custom interpolated message is rendered on the wrong-exception-type failure branch.

**Comment by Evangelink** [comment]
### 🧪 Test quality grade — PR #9532

11 test methods graded across 3 files. All score in the A band. The new tests form a tightly paired, consistent set covering success paths, failure-message rendering, and edge-case boundary conditions for the new interpolated-string-handler overloads. No structural or assertion-strength issues found.

<table>
  <thead>
    <tr><th>Δ</th><th>Test</th><th>Grade</th><th>Band</th><th>Notes</th></tr>
  </thead>
  <tbody>
    <tr>
      <td>new</td>
      <td><code>AssertTests.<br>ThrowsAsync_<br>WithInterpolation_<br>DoesNotEvaluateMessageOnSuccess</code></td>
      <td>A</td>
      <td>90–100</td>
      <td>Sentinel bool confirms lazy interpolation evaluation; clear AAA with a meaningful behavioral assertion.</td>
    </tr>
    <tr>
      <td>new</td>
      <td><code>AssertTests.<br>ThrowsAsync_<br>WithInterpolation_<br>RendersMessageOnFailure</code></td>
      <td>A</td>
      <td>90–100</td>
      <td>Exact multi-line message match verifies the rendered interpolation on the no-exception path.</td>
    </tr>
    <tr>
      <td>new</td>
      <td><code>AssertTests.<br>ThrowsAsync_<br>WithInterpolation_<br>RendersMessageOnWrongExceptionType</code></td>
      <td>A</td>
      <td>90–100</td>
      <td>Verifies the rendered failure message for the wrong-exception-type path; well-focused.</td>
    </tr>
    <tr>
      <td>new</td>
      <td><code>AssertTests.<br>ThrowsExactlyAsync_<br>WithInterpolation_<br>DoesNotEvaluateMessageOnSuccess</code></td>
      <td>A</td>
      <td>90–100</td>
      <td>Mirrors the ThrowsAsync lazy-eval sibling; sentinel bool pattern is effective.</td>
    </tr>
    <tr>
      <td>new</td>
      <td><code>AssertTests.<br>ThrowsExactlyAsync_<br>WithInterpolation_<br>RendersMessageOnFailure</code></td>
      <td>A</td>
      <td>90–100</td>
      <td>Exact message match on the ThrowsExactlyAsync no-exception failure path.</td>
    </tr>
    <tr>
      <td>new</td>
      <td><code>AssertTests.<br>ThrowsExactlyAsync_<br>WithInterpolation_<br>RendersMessageOnWrongExceptionType</code></td>
      <td>A</td>
      <td>90–100</td>
      <td>Verifies the "exact type" vs "or derived" distinction is preserved in the failure message.</td>
    </tr>
    <tr>
      <td>mod</td>
      <td><code>PreferAsyncAssertionAnalyzerTests.<br>WhenAssertionUsesInterpolatedStringHandlerOverload_<br>CodeFixUsesAsyncAssertion</code></td>
      <td>A</td>
      <td>90–100</td>
      <td>VerifyCodeFixAsync validates both the diagnostic and the full transformation; well-structured analyzer test.</td>
    </tr>
    <tr>
      <td>new</td>
      <td><code>PreferAsyncAssertionAnalyzerTests.<br>WhenAssertThrowsUsesInterpolatedStringHandlerOverload_<br>CodeFixUsesAsyncAssertion</code></td>
      <td>A</td>
      <td>90–100</td>
      <td>Clean parallel to the ThrowsExactly sibling; VerifyCodeFixAsync verifies both detection and fix.</td>
    </tr>
    <tr>
      <td>new</td>
      <td><code>TestMethodValidatorTests.<br>IsValidTestMethodShouldReturnFalseAndReportTargetedWarningForGenericValueTaskReturnType</code></td>
      <td>A</td>
      <td>90–100</td>
      <td>Three-layered assertions (result, count, content) thoroughly cover the targeted-warning contract.</td>
    </tr>
    <tr>
      <td>new</td>
      <td><code>TestMethodValidatorTests.<br>IsValidTestMethodShouldReportGenericSignatureWarningWhenGenericValueTaskMethodIsAlsoInvalidForAnotherReason</code></td>
      <td>A</td>
      <td>90–100</td>
      <td>Negative assertion guards the over-eager targeted message; inline comment clarifies intent precisely.</td>
    </tr>
    <tr>
      <td>new</td>
      <td><code>TestMethodValidatorTests.<br>IsValidTestMethodShouldReturnTrueForMethodsWithValueTaskReturnType</code></td>
      <td>A</td>
      <td>90–100</td>
      <td>Verifies both the true result and absence of warnings; clear and focused.</td>
    </tr>
  </tbody>
</table>

<sub>This advisory comment was generated automatically. Grades are heuristic
and informational — they do not block merging. Re-run with
`/grade-tests`.</sub>



> 🤖 **Automated content by GitHub Copilot.** Posted via a maintainer's GitHub token, so it appears under their account — the account owner did **not** write or approve this content personally. Generated by the [Grade Tests on PR (on open / sync)](https://github.com/microsoft/testfx/actions/runs/28433205773/agentic_workflow) workflow. · 524.4 AIC · ⌖ 13.1 AIC · ⊞ 43.7K · [◷]( · [◷](https://github.com/search?q=repo%3Amicrosoft%2Ftestfx+%22gh-aw-workflow-call-id%3A+microsoft%2Ftestfx%2Fgrade-tests-on-pr.agent%22&type=issues))

File: src/TestFramework/TestFramework/Assertions/Assert.ThrowsException.InterpolatedStringHandlers.cs (hunk 2 above)
Comment (Evangelink) [reply src/TestFramework/TestFramework/Assertions/Assert.ThrowsException.InterpolatedStringHandlers.cs]: Addressed in 8db561ec7 as well: the `ThrowsExactlyAsync` ISH constructor now uses the same `Task.Run(() => IsThrowsAsyncFailingAsync(...))` guard, so it's no longer susceptible to the SynchronizationContext deadlock.

File: test/UnitTests/TestFramework.UnitTests/Assertions/AssertTests.ThrowsExceptionTests.cs (hunk 5 above)
Comment (Evangelink) [reply test/UnitTests/TestFramework.UnitTests/Assertions/AssertTests.ThrowsExceptionTests.cs:363]: Added in 8db561ec7: `ThrowsAsync_WithInterpolation_RendersMessageOnWrongExceptionType` (plus a `ThrowsExactlyAsync` variant) covers the wrong-exception-type failure branch and asserts the interpolated message is rendered. (Note: my matching reply earlier accidentally landed on the null-safety thread — this is the correct location.)

File: test/UnitTests/MSTestAdapter.PlatformServices.UnitTests/Discovery/TestMethodValidatorTests.cs (hunk 4 above)
Comment (Evangelink) [reply test/UnitTests/MSTestAdapter.PlatformServices.UnitTests/Discovery/TestMethodValidatorTests.cs:154]: Thanks for confirming — no change needed; leaving the `TestContainer` convention as-is. Resolving.

File: src/Adapter/MSTestAdapter.PlatformServices/Discovery/TestMethodValidator.cs (hunk 6 above)
Comment (Evangelink) [reply src/Adapter/MSTestAdapter.PlatformServices/Discovery/TestMethodValidator.cs:94]: Confirmed correct — name comparison is null-safe and the CLR literal is canonical, so no change needed here. (My earlier reply about the wrong-exception test was meant for the test-scenario thread; correcting the record.) Resolving.

**Review by 0101 — APPROVED** [review]
